e3/electrolyt/run.py

```diff
@@ -29,6 +34,97 @@ def run(self):
                 if hasattr(self, method_name):
                     getattr(self, method_name)()
 
+    def do_build(self):
+        """Run the build primitive after setting up the sandbox."""
+        self.data.data.sandbox = self.sandbox
+        anod_driver = AnodDriver(anod_instance=self.data.data, store=self.store)
+        anod_driver.activate()
+        anod_driver.anod_instance.build_space.create(quiet=True)
+        if not getattr(anod_driver.anod_instance, 'build', False):
+            logger.error('primitive build not implemented')
+            self.status = STATUS.failure
+            return
+        anod_driver.anod_instance.build()
+        self.status = STATUS.success
+
+    def do_test(self):
+        """Run the test primitive."""
+        self.data.data.sandbox = self.sandbox
+        anod_driver = AnodDriver(anod_instance=self.data.data, store=self.store)
+        anod_driver.activate()
+        anod_driver.anod_instance.build_space.create(quiet=True)
+        if not getattr(anod_driver.anod_instance, 'test', False):
+            logger.error('primitive test not implemented')
+            self.status = STATUS.failure
+            return
+        anod_driver.anod_instance.test()
+        self.status = STATUS.success
+
+    def do_checkout(self):
+        """Get sources from vcs to sandbox vcs_dir."""
+        repo_name = self.data.data[0]
+        repo_url = self.data.data[1]['url']
+        repo_revision = self.data.data[1]['revision']
+        repo_vcs = self.data.data[1]['vcs']
+        if repo_vcs != 'git':
+            logger.error('%s vcs type not supported' % repo_vcs)
+            self.status = STATUS.failure
+            return
+        repo_dir = os.path.join(self.sandbox.vcs_dir, repo_name)
+        g = GitRepository(repo_dir)
+        if e3.log.default_output_stream is not None:
+            g.log_stream = e3.log.default_output_stream
+        g.init()
+        g.update(repo_url, repo_revision, force=True)
+        self.status = STATUS.success
+
+    def do_createsource(self):
+        """Prepare src from vcs to cache using sourcebuilders."""
+        source_name = self.data.source_name
+        tmp_cache_dir = os.path.join(self.sandbox.tmp_dir, 'cache')
+        src = self.sandbox.vcs_dir
+        for src_builder in self.data.spec.source_pkg_build:
+            if src_builder.name == source_name:
+                repo_dict = {}
+                src_dir = os.path.join(src, src_builder.checkout[0])
+                dest_dir = os.path.join(tmp_cache_dir, source_name)
+                repo_dict[source_name] = {"working_dir": src_dir}
```

Comment:
you're missing the repo state, if you want to do that in another commit please add:

  ??? missing repository state

and add an issue